### PR TITLE
Create a Provider API script template

### LINKED ISSUE #92: [Feature] Add a Provider API Script template
## Problem Description

Previously, we identified writing the scripts that pull data from previously unknown APIs (Provider API Scripts) as the best entry points for new community contributors.

The Provider API Scripts are required to provide either a **license url** or a **pair of license name and license version**. If those were not valid, the ImageStore repository class would silently just not write the data for the image. We did not validate the licenses in the Provider API scripts because we felt that it would add complexity of learning about inner details of license naming for the new contributors. 

During the discussion of moving license validation from the ImageStore repository to the individual provider API scripts, we probably found a way to both validate the licenses and clearly show if validation failed, as well as not to bog the new contributors down with details about licenses (there are more than 500 versions!) 

## Solution Description

We can create a [cookiecutter](https://cookiecutter.readthedocs.io/en/1.7.2/)-like template for creating a Provider API script:
Add a folder with a **Provider API Script** Template, a **Provider API Script Test** template, and a **Provider API Workflow** Template
Write a CLI script that fills in the templates with Provider name, and save the files in the appropriate locations.

To make sure that license validation is simplified for the new contributors, we can: 
1. provide clearer documentation about what we can accept as a valid license
2. import the function that validates license_name and license_version pair into the Provider API Script template so that the user could use it during script writing and testing.
3. have a clearer indication if the license_url or license_name and license_version pair are invalid so that the user can fix them if possible. Currently, if the license data is not valid, the scripts do not provide very clear feedback, failing quietly.

## Alternatives

- Move license validation to the Provider API scripts, and add more documentation about it.
This would probably make the extraction and saving of license information easier. However, with a template, we can also get better-quality scripts by providing best practices and good examples within it.

## Additional context

This template will also help when we are adding Audio sources in the short term, and other Open licenses in the long term.

### PR DESCRIPTION
**Pull request by obulat**
Fixes #92 

This PR creates a Cookiecutter-like script for writing Provider API scripts.
It needs more testing and documentation.

To run the script, create and activate a virtual environment, navigate to the templates folder and run it with provider name as a parameter:
```
cd src/cc_catalog_airflow/templates/
python3 create_api_script.py <provider_name>
```

This will create three files:
1. The API provider script with a skeleton implementation
2. The test file (blank)
3. The workflow file which doesn't really need much change, I suppose.

There are some instructions and a lot of #TODO comments in the file. It would be really helpful if you note any problems, difficulties, or anything that's unclear during testing.

Signed-off-by: Olga Bulat <obulat@gmail.com>

### CODE REVIEW

File: src/cc_catalog_airflow/templates/template_provider.py_template [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+LIMIT = 0  # number of items per page in API response
+DELAY = 1  # in seconds
+RETRIES = 3
+HOST = '{{host URL}}'
+ENDPOINT = f'https://{HOST}/{{API PATH}}'
+# TODO: Add Provider constant to the
+#  `src/cc_catalog_airflow/dags/util/loader/provider_details.py` file
+PROVIDER = prov.{provider_upper_case}_DEFAULT_PROVIDER
+# TODO: Add the API key to `src/cc_catalog_airflow/env.template`
+#  Do not add your API key value!
```
Comment (zackkrida) [inline src/cc_catalog_airflow/templates/template_provider.py_template:58]:
```suggestion
#  Do not hardcode your API key value!
```

**Review by zackkrida — APPROVED** [review]
Added some small changes, I'm sure this will be reviewed and re-reviewed a few times until we finish. You might also be interested this tool:

https://plopjs.com/

File: src/cc_catalog_airflow/templates/template_provider.py_template [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+
+Try to write small functions that are easier to test. Don't forget to
+write tests, too!
+
+You can test your script during development by running it:
+`python -m <your_script_name>.py`
+
+To extract information from html, you can use lxml.html.
+If you need to look at the examples, the best example for dated script is
+`wikimedia_commons.py`, for not dated: `science_museum.py`.
```
[the remark this answers is not in the document]
Comment (obulat) [reply src/cc_catalog_airflow/templates/template_provider.py_template:35]: Oh, I didn't think of that meaning of the word "dated" :)
For Flickr and Wikimedia, we only ingest the latest data, not all of it from the start. So, we need to set the date from which to ingest. That date is what I refer to as "dated"

[the remark this answers is not in the document]
Comment (zackkrida) [reply src/cc_catalog_airflow/templates/template_provider.py_template:35]: Thanks for explaining! I'll try to think of clearer language 🤔 
"Dated" is commonly a shorthand for "out of date / expired / old fashioned" in English, which is really weird now that I think about it.

Comment (zackkrida) [inline src/cc_catalog_airflow/templates/template_provider.py_template:35]:
```suggestion
Recommended examples:
- For scripts requiring a start date: `wikimedia_commons.py`
- For scripts without a start date: `science_museum.py`.
```

maybe something like this?

**Comment by obulat** [comment]
> Added some small changes, I'm sure this will be reviewed and re-reviewed a few times until we finish. You might also be interested this tool:
> 
> https://plopjs.com/

This tool looks neat!

What I think is good about the current approach is the fact that the contributors will not have to install any additional packages, only the ones required to run the API script(`requests`, `lxml` maybe?), and they won't even have to set up Catalog in Docker! On Windows, it's very difficult for people new to programming: you have to either have Professional edition of Windows, or install WSL2 system.

**Comment by zackkrida** [comment]
Yes, it's great that they're generally small standalone scripts that are straightforward to review and test. This is looking great and I'd like @dhruvkb and @krysal to take a look this week too

**Comment by zackkrida** [comment]
@krysal it would be excellent for you to test this template by writing a Provider API Script for stocksnap.io. I have gotten more information that should help.

They have an api we can use available at [https://stocksnap.io/api/load-photos/date/desc/1](https://stocksnap.io/api/load-photos/date/desc/1) where "1" is the page number. I did some simple testing and there's around 33k records at this API. Please copy Olga's template files into a new PR and document the experience of using them. This should be a great way to learn what is missing in the files; any questions you have or problems you encounter are a _good thing_ and will help us make the template better!

**Comment by obulat** [comment]
Quick docs (to be expanded!):
To create an API script using, run 
```
cd src/cc_catalog_airflow/templates/
python3 create_api_script.py --provider <name>
```
By default, it creates a script for images. If you are writing a script for an audio provider, you need to add `--media audio` argument:
```
python3 create_api_script.py --provider <name> --media audio
```

File: src/cc_catalog_airflow/templates/create_api_script.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
+import argparse
+from pathlib import Path
+
+
+IMAGE_STORE_INIT = 'image_store = ImageStore(provider=PROVIDER)'
+AUDIO_STORE_INIT = 'audio_store = AudioStore(provider=PROVIDER)'
+
+
+def _get_filled_template(template_path, provider, media_type=None):
```
Comment (zackkrida) [inline src/cc_catalog_airflow/templates/create_api_script.py:9]:
```suggestion
def _get_filled_template(template_path, provider, media_type='image'):
```

**Review by zackkrida — CHANGES_REQUESTED** [review]
> When attempting to run the new `create_api_script` command I got the following errors:
> 
> ```shell
> FileNotFoundError: [Errno 2] No such file or directory: 'dags/provider_api_scripts/stocksnap.py'
> ```

Fixed by switching from dot relative path (Path(`'.'`)) to absolute (`Path(__file__)`)

**Comment by obulat** [comment]
TODO: Add a TODO to add a popularity metric to meta_data (views/listens, likes, downloads etc).

File: src/cc_catalog_airflow/templates/template_provider.py_template [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+
+"""
+This is template for an API script. Broadly, there are several steps:
+1. Download batches of information for the query for openly-licensed media
+2. For each item in batch, extract the necessary meta data.
+3. Save the metadata using ImageStore.add_item or AudioStore.add_item methods
+
+Try to write small functions that are easier to test. Don't forget to
+write tests, too!
+
```
Comment (krysal) [inline src/cc_catalog_airflow/templates/template_provider.py_template:29]: Good overview to add a source, and remarking tests 👍

**Review by krysal — CHANGES_REQUESTED** [review]
Sorry the delayed review! I was able to run the `create_api_script.py` only with its full path. Otherwise the `No such file or directory` error persists. I had to do some tweaks but I agree this will be a very good starting point for new contributors once finished.

File: src/cc_catalog_airflow/templates/workflow.py_template [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+
+DAG_DEFAULT_ARGS = {
+    'owner': 'data-eng-admin',
+    'depends_on_past': False,
+    'start_date': datetime(2020, 1, 15),
+    'email_on_retry': False,
+    'retries': 3,
+    'retry_delay': timedelta(days=1),
+}
```
Comment (krysal) [inline src/cc_catalog_airflow/templates/workflow.py_template:17]: The [dag_factory][dag_factory_link] already use almost the same default args as here, from a [config][config_link] file. The workflow file can be slightly smaller and repeat less code if we use the factory, as the workflow script for the Smithsonian Institute and Flickr for example. Have you considered this?

[dag_factory_link]: https://github.com/WordPress/openverse-catalog/blob/template/src/cc_catalog_airflow/dags/util/dag_factory.py
[config_link]: https://github.com/WordPress/openverse-catalog/blob/template/src/cc_catalog_airflow/dags/util/config.py#L7:L14

Comment (obulat) [reply src/cc_catalog_airflow/templates/workflow.py_template:17]: No, I haven't looked into the dag_factory, thank you for the suggestion, I will look into it!

File: src/cc_catalog_airflow/templates/template_provider.py_template [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+        'creator_url': creator_url,
+        'foreign_identifier': foreign_identifier,
+        'foreign_landing_url': foreign_landing_url,
+        'image_url': image_url,
+        'height': height,
+        'width': width,
+        'thumbnail_url': thumbnail,
+        'item_license': item_license,
+        'metadata': metadata,
+        'tags': tags
```
Comment (krysal) [inline src/cc_catalog_airflow/templates/template_provider.py_template:254]:
```suggestion
        'license_': item_license.license,
        'license_version': item_license.version,
        'meta_data': metadata,
        'raw_tags': tags
```
Some name corrections

File: src/cc_catalog_airflow/templates/template_provider.py_template [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+    # audio_url, duration = _get_audio_info(media_data)
+    image_url, height, width = _get_image_info(media_data)
+    if image_url is None:
+        print (f"Found no media url:")
+        print(f"{json.dumps(media_data, indent=2)}")
+        check_and_save_json_for_test('no_image_url', media_data)
+        return None
+    item_license = _get_license(media_data)
+    if item_license is None:
+        print (f"Found no item license:")
```
Comment (krysal) [inline src/cc_catalog_airflow/templates/template_provider.py_template:230]:
```suggestion
        print(f"Found no foreign landing url:")
        print(f"{json.dumps(media_data, indent=2)}")
        check_and_save_json_for_test('no_foreign_landing_url', media_data)
        return None
    # audio_url, duration = _get_audio_info(media_data)
    image_url, height, width = _get_image_info(media_data)
    if image_url is None:
        print(f"Found no media url:")
        print(f"{json.dumps(media_data, indent=2)}")
        check_and_save_json_for_test('no_image_url', media_data)
        return None
    item_license = _get_license(media_data)
    if item_license is None:
        print(f"Found no item license:")
```
Extra spaces

File: src/cc_catalog_airflow/templates/template_provider.py_template [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+    metadata = {}
+    some_other_key_value = item.get('some_other_key')
+    if some_other_key_value is not None:
+        metadata['some_other_key'] = some_other_key_value
+    return metadata
+
+
+def _get_tags(item):
+    """What tags do we save???"""
+    return item.get('tags')
```
Comment (krysal) [inline src/cc_catalog_airflow/templates/template_provider.py_template:313]: I see these tags (`raw_tags`) as keywords that describe or are related to the media item in the provider view. E.g. StockSnap have both, tags and keywords.

**Review by krysal — COMMENTED** [review]
I'm almost done reviewing, found some little details.

**Review by zackkrida — APPROVED** [review]
It works in any directory now :)

**Review by zackkrida — DISMISSED** [review]

**Review by zackkrida — DISMISSED** [review]

**Review by krysal — APPROVED** [review]
I finished _testing the test template_ and made some minor adjustments, now it's ready to go! 🎶

**Review by zackkrida — APPROVED** [review]
LGTM!

**Review by zackkrida — APPROVED** [review]

**Review by krysal — APPROVED** [review]
This is really helpful, thank you! 🌟

**Pull request merged**